### PR TITLE
storage: Be multipath safe.

### PR DESCRIPTION
**Pull request by mvollmer**
Show the whole picture about multipath devices and don't let people
mess with improperly configured ones.

### CODE REVIEW

**Comment by mvollmer** [comment]
Next up: integration tests.

**Comment by mvollmer** [comment]
> Next up: integration tests.

Done.

File: pkg/storaged/client.js [hunk 1]
```diff
@@ -153,13 +153,42 @@ define([
...
-                client.drives_block[block.Drive] = block;
+            if (!client.blocks_part[path] && client.drives_multipath_blocks[block.Drive] !== undefined)
+                client.drives_multipath_blocks[block.Drive].push(block);
+        }
+        for (path in client.drives_multipath_blocks) {
+            var all_blocks = client.drives_multipath_blocks[path];
+            var mpath_members = [ ];
+            var non_mpath_members = [ ];
+
+            all_blocks.sort(function (a, b) { return a.DeviceNumber - b.DeviceNumber; });
```
Comment (dperpeet) [inline pkg/storaged/client.js]: why do you use the difference between the device numbers as the sort key?

File: pkg/storaged/client.js [hunk 2]
```diff
@@ -153,13 +153,42 @@ define([
...
+                    mpath_members.push(all_blocks[i]);
+                else
+                    non_mpath_members.push(all_blocks[i]);
+            }
+
+            /* A valid multipath drive has exactly one block device
+             * that is not a "mpath_member".
+             */
+
+            if (non_mpath_members.length == 1) {
```
Comment (dperpeet) [inline pkg/storaged/client.js]: can we use `===` operator here?

File: test/guest/cockpit-rhel-7.setup [hunk 3]
```diff
@@ -48,6 +48,7 @@ selinux-policy-targeted \
 subscription-manager \
 storaged \
 storaged-lvm2 \
+device-mapper-multipath \
```
Comment (dperpeet) [inline test/guest/cockpit-rhel-7.setup]: the other entries in this list appear to be in alphabetical order...

File: pkg/storaged/client.js (hunk 1 above)
Comment (mvollmer) [inline pkg/storaged/client.js]: Yes, agreed.  This is a common pattern in the storage code.  I'll factor it out and document it.

File: pkg/storaged/client.js (hunk 2 above)
Comment (mvollmer) [inline pkg/storaged/client.js]: Yes, but why?

File: test/guest/cockpit-rhel-7.setup (hunk 3 above)
Comment (mvollmer) [inline test/guest/cockpit-rhel-7.setup]: good catch

**Comment by dperpeet** [comment]
code looks good, need to take this for a spin

**Comment by dperpeet** [comment]
trying this out looks good
for a follow-up, it'd be nice to get a note why there are no options to actually use a multipathed disk
(enable multipathd)

**Pull request closed without merging**